### ISSUE TITLE
CSE on agent pool node fails when configuring azure-cni-networkmonitor

### ISSUE THREAD

**Issue by mhyllander** [Azure/acs-engine#3638]
**Is this a request for help?**: Maybe

---

**Is this an ISSUE or FEATURE REQUEST?** (choose one): ISSUE

---

**What version of acs-engine?**: v0.19.0 - v0.20.6

---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**: kubernetes 1.11.0, 1.11.1

**What happened**:
The CSE that configures the agent pool node(s) fails with exit code 4.

**What you expected to happen**:
Cluster deployment succeeds.

**How to reproduce it** (as minimally and precisely as possible):
When acs-engine v0.19.0 was released with support for kubernetes 1.11.0, I tried to deploy a cluster but it failed. The CSE that configured the agent pool with availabilityProfile="VirtualMachineScaleSets" failed with exit code 4.

Now I have tested acs-engine v0.20.6 with kubernetes 1.11.1 with the same result. Then I  tried deploying the agent pool with availabilityProfile="AvailabilitySet" so I could log in to the node where the CSE fails and look at CSE log file (/var/log/azure/cluster-provision.log).

I found that the problem is this line: https://github.com/Azure/acs-engine/blob/master/parts/k8s/kubernetesagentcustomdata.yml#L132

The sed command fails because /etc/kubernetes/addons/azure-cni-networkmonitor.yaml does not exist.

I have tested commenting out that line and compiling acs-engine from source. With this patched version of acs-engine, deployment succeeds.

**Anything else we need to know**:
I'm deploying using a custom VNet.

**Comment by CecileRobertMichon** [comment]
@mhyllander thanks for reporting! Could you please share the last few lines from /var/log/azure/cluster-provision.log?

**Comment by mhyllander** [comment]
@CecileRobertMichon I've reproduced the error, using acs-engine v0.20.7.

The final lines of /var/log/azure/cluster-provision.log on the agent pool node are:

```
+ for i in '$(seq 1 $retries)'
+ timeout 30 systemctl daemon-reload
+ timeout 30 systemctl restart kubelet
Job for kubelet.service failed because the control process exited with error code. See "systemctl status kubelet.service" and "journalctl -xe" for details.
+ '[' 1 -eq 0 ']'
+ '[' 100 -eq 100 ']'
+ return 1
+ RESTART_STATUS=1
+ systemctl status kubelet --no-pager -l
+ '[' 1 -ne 0 ']'
+ echo 'kubelet could not be started'
kubelet could not be started
+ exit 4
```

``` bash
# systemctl status kubelet.service
● kubelet.service - Kubelet
   Loaded: loaded (/etc/systemd/system/kubelet.service; disabled; vendor preset: enabled)
   Active: inactive (dead)

Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 systemd[1]: Starting Kubelet...
Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 bash[24669]: sed: can't read /etc/kubernetes/addons/azure-cni-networkmonitor.yaml: No such file or directory
Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 systemd[1]: kubelet.service: Control process exited, code=exited status=2
Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 systemd[1]: Failed to start Kubelet.
Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 systemd[1]: kubelet.service: Unit entered failed state.
Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 systemd[1]: kubelet.service: Failed with result 'exit-code'.
Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 systemd[1]: kubelet.service: Service hold-off time over, scheduling restart.
Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 systemd[1]: Stopped Kubelet.
Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 systemd[1]: kubelet.service: Start request repeated too quickly.
Aug 14 11:16:36 k8s-agentpool1-28837412-vmss000000 systemd[1]: Failed to start Kubelet.
```

``` bash
# cat /etc/systemd/system/kubelet.service
[Unit]
Description=Kubelet
ConditionPathExists=/usr/local/bin/kubelet
# Required

[Service]
Restart=always
EnvironmentFile=/etc/default/kubelet
SuccessExitStatus=143
ExecStartPre=/bin/bash /opt/azure/containers/kubelet.sh
ExecStartPre=/bin/mkdir -p /var/lib/kubelet
ExecStartPre=/bin/mkdir -p /var/lib/cni
ExecStartPre=/bin/bash -c "if [ $(mount | grep \"/var/lib/kubelet\" | wc -l) -le 0 ] ; then /bin/mount --bind /var/lib/kubelet /var/lib/kubelet ; fi"
ExecStartPre=/bin/mount --make-shared /var/lib/kubelet
# This is a partial workaround to this upstream Kubernetes issue:
#  https://github.com/kubernetes/kubernetes/issues/41916#issuecomment-312428731
ExecStartPre=/sbin/sysctl -w net.ipv4.tcp_retries2=8
ExecStartPre=-/sbin/ebtables -t nat --list
ExecStartPre=-/sbin/iptables -t nat --list
ExecStart=/usr/local/bin/kubelet \
        --enable-server \
        --node-labels="${KUBELET_NODE_LABELS}" \
        --v=2 \
        --volume-plugin-dir=/etc/kubernetes/volumeplugins \
        $KUBELET_CONFIG $KUBELET_OPTS \
        $KUBELET_REGISTER_NODE $KUBELET_REGISTER_WITH_TAINTS

[Install]
```

``` bash
# cat /opt/azure/containers/kubelet.sh
#!/bin/bash

# SNAT outbound traffic from pods to destinations outside of VNET.
iptables -t nat -A POSTROUTING -m iprange ! --dst-range 168.63.129.16 -m addrtype ! --dst-type local ! -d 10.0.0.0/8 -j MASQUERADE
sed -i "s|<azureCNINetworkMonitorImage>|containernetworking/networkmonitor:v0.0.4|g" "/etc/kubernetes/addons/azure-cni-networkmonitor.yaml"
```

``` bash
# ls -l /etc/kubernetes/addons/
ls: cannot access '/etc/kubernetes/addons/': No such file or directory
```

As you see the sed command is trying to modify /etc/kubernetes/addons/azure-cni-networkmonitor.yaml, but the /etc/kubernetes/addons/ directory doesn't exist on the agent pool node.

I think the problem is that the azure-cni-networkmonitor addon is only installed on the master, so the sed command in /opt/azure/containers/kubelet.sh should also only run on the master.

In other words, this line is OK:
  https://github.com/Azure/acs-engine/blob/master/parts/k8s/kubernetesmastercustomdata.yml#L208
But this line should be removed?
  https://github.com/Azure/acs-engine/blob/master/parts/k8s/kubernetesagentcustomdata.yml#L139

**Comment by mhyllander** [comment]
I have a question as well: is there a reason why azure-cni-networkmonitor is deployed as a DaemonSet on worker nodes only? Why isn't it running on the master nodes as well, since I assume they too have network configuration that might need cleaning up?

**Comment by CecileRobertMichon** [comment]
@jackfrancis this might be a bug from https://github.com/Azure/acs-engine/pull/3198

**Comment by jackfrancis** [comment]
@mhyllander It's definitely a bug that there's a reference to the addons directory in the agent kubelet.sh

**Comment by jackfrancis** [comment]
@mhyllander In theory, you're right that the value of that daemonset is applicable on master nodes, but by default acs-engine-built clusters don't ship w/ schedulable masters. Will address as well.
